cvat/apps/dataset_manager/formats/coco.py

```diff
@@ -32,8 +33,16 @@ def _import(src_file, temp_dir, instance_data, load_data_callback=None, **kwargs
             load_data_callback(dataset, instance_data)
         import_dm_annotations(dataset, instance_data)
     else:
+        rename_mapping, src_file = update_annotation_data(src_file)
+        #To get reverse map with image names without extension
+        reverse_mapping = {value: key for key, value in rename_mapping.items()}
+        reverse_no_ext_mapping = {key.rsplit('.', 1)[0]: value.rsplit('.', 1)[0] for key, value in reverse_mapping.items()}
+
         dataset = Dataset.import_from(src_file.name,
             'coco_instances', env=dm_env)
+        #Reverse mapping names
+        for item in dataset:
+            item.id = reverse_no_ext_mapping[item.id]
```

Comment:
@zhiltsov-max 

I've been experimenting with the implementation provided in [#7523(comment)](https://github.com/opencv/cvat/issues/7523#issuecomment-1988060365), and I encountered a circular import error while attempting to load the rename_mapping variable into bindings.py.
  
`ImportError: cannot import name 'GetCVATDataExtractor' from partially initialized module 'cvat.apps.dataset_manager.bindings' (most likely due to a circular import) (/home/adkbbx/projects/CVAT/cvat/cvat/apps/dataset_manager/bindings.py)` 

Which I thought would be the way of implementing this logic please do correct me if I was wrong with the implementation.

On furthur debugging I found that the 

`self._frame_mapping` variable inside the [bindings.py](https://github.com/opencv/cvat/blob/develop/cvat/apps/dataset_manager/bindings.py#L1276C40-L1276C54) only contains the value `{"image1": 1}` may be since same keys were used since path "Image1" was same for both the images in the dataset and only the value of fram got update from 0->1 which I felt would be the cause of  multiple annotations being placed on the same image on my current implementation.

To resolve this, I attempted to updating the _frame_mapping attribute with the instance_data and this seems to resolve the issue of multiple annotations being placed on the same image on my current implementation but while I was testing the export of datasets the annotations aren't getting updated into json file.
`"annotations": [] ` 

So I think updating _frame_mapping might be the cause of this export error which I am unable to debug.

I would appreciate any hints, insights or suggestions for the implementation of this solution further. I'm sorry for the frequent contact and thank you for your time and patience! 🙇

Reply:
> I've been experimenting with the implementation provided in https://github.com/opencv/cvat/issues/7523#issuecomment-1988060365, and I encountered a circular import error while attempting to load the rename_mapping variable into bindings.py.

>> ImportError: cannot import name 'GetCVATDataExtractor' from partially initialized module 'cvat.apps.dataset_manager.bindings' (most likely due to a circular import) (/home/adkbbx/projects/CVAT/cvat/cvat/apps/dataset_manager/bindings.py)

> Which I thought would be the way of implementing this logic please do correct me if I was wrong with the implementation.

The suggestion in the comment was to add a new function parameter to `import_dm_annotations()` in `bindings.py`. I believe, this doesn't require importing any modules. Then, you can supply the constructed mapping into this function, when it's called in `coco.py`. After this, in `import_dm_annotations()`, the `match_dm_item()` gets called. It also needs a new parameter. `match_dm_item()` needs to be updated to use this new parameter to map the item name back into the original one and find the matching frame id in the job or task.

Reply:
@zhiltsov-max 
I've pushed a new commit attempting to address the issue and think I was able to incorporated the suggestions you provided earlier, including adding a new parameter to import_dm_annotations() in bindings.py and updating match_dm_item() accordingly. Despite these changes,  It seems that the annotations are still being duplicated on some images even though the anotations are being accepted now. As of what I am able to understand from my debugging the way `instance_dataset._frame_mapping` updates when there are images with same name and different extensions might be the cause of this issue. 
Any further guidance you can provide would be greatly appreciated. Thank you for your continued support and feedback.

Reply:
That's very nice to see advances in solving this problem.

> As of what I am able to understand from my debugging the way `instance_dataset._frame_mapping` updates when there are images with same name and different extensions might be the cause of this issue.

Yeah, it seems to be the issue now. From what I've checked, we can try to change what's stored in these frame mappings, as they seem to only be used for frame matching. Probably, we can include file extensions in these mappings as well.

Reply:
@zhiltsov-max 
I have tried to append extensions to the _frame_mappings but I am still unable to solve the issue of multiple annotations being placed on the same image, I have also pushed the code for your reference. Do let me know if you find anything that might be causing this issue.
I will continue working on this tommorow as soon as I am awake. Thanks to your support I am really able to learn alot solving this issue!

Reply:
@zhiltsov-max 
I'm sorry for the delay in my response. I was just trying to debug the problem and found another issue which is the source of multiple annotations being placed on the same image.

So currently I was importing the  annotations exported from CVAT website after annotating the images with same names and apparently the image id's of same image names and different extensions remain same while exporting the annotations. 

```json
{
    "licenses": [
        {
            "name": "",
            "id": 0,
            "url": ""
        }
    ],
    "info": {
        "contributor": "",
        "date_created": "",
        "description": "",
        "url": "",
        "version": "",
        "year": ""
    },
    "categories": [
        {
            "id": 1,
            "name": "person",
            "supercategory": ""
        }
    ],
    "images": [
        {
            "id": 1,
            "width": 360,
            "height": 360,
            "file_name": "image1.gif",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        },
        {
            "id": 1,
            "width": 360,
            "height": 360,
            "file_name": "image1.jpg",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        },
        {
            "id": 1,
            "width": 1200,
            "height": 1200,
            "file_name": "image1.png",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        },
        {
            "id": 2,
            "width": 360,
            "height": 360,
            "file_name": "image2.jpg",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        },
        {
            "id": 3,
            "width": 360,
            "height": 360,
            "file_name": "image3.jpg",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        },
        {
            "id": 3,
            "width": 1200,
            "height": 1200,
            "file_name": "image3.png",
            "license": 0,
            "flickr_url": "",
            "coco_url": "",
            "date_captured": 0
        }
    ],
    "annotations": [
        {
            "id": 1,
            "image_id": 1,
            "category_id": 1,
            "segmentation": [],
            "area": 28087.773,
            "bbox": [
                131.64,
                32.79,
                95.83,
                293.1
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        },
        {
            "id": 2,
            "image_id": 1,
            "category_id": 1,
            "segmentation": [],
            "area": 31587.525000000005,
            "bbox": [
                128.62,
                29.34,
                105.75,
                298.7
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        },
        {
            "id": 3,
            "image_id": 1,
            "category_id": 1,
            "segmentation": [],
            "area": 371765.10079999996,
            "bbox": [
                415.83,
                130.94,
                395.68,
                939.56
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        },
        {
            "id": 4,
            "image_id": 2,
            "category_id": 1,
            "segmentation": [],
            "area": 32492.586,
            "bbox": [
                123.44,
                31.07,
                108.78,
                298.7
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        },
        {
            "id": 5,
            "image_id": 3,
            "category_id": 1,
            "segmentation": [],
            "area": 36013.08240000001,
            "bbox": [
                121.71,
                26.32,
                117.84,
                305.61
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        },
        {
            "id": 6,
            "image_id": 3,
            "category_id": 1,
            "segmentation": [],
            "area": 351750.0474,
            "bbox": [
                405.76,
                133.81,
                372.66,
                943.89
            ],
            "iscrowd": 0,
            "attributes": {
                "occluded": false,
                "rotation": 0.0
            }
        }
    ]
}
```

Therefore importing the same annotations was the source of this issue. Therefore I just updated the image id's and it seems like the importing of the annotations have no issue now after the above fix but while exporting the annotations the annotation with same image names but different extensions have the same image id's due to which this behaviour is still seen on importing the CVAT exported COCO dataset. I am currently going through the COCO format annotation exporter code to debug this issue. Do let me know if you have any inputs for me!

Reply:
Hi, any updates? Please add a test to cover the changes made and this PR can be merged. Please follow the instructions [here](https://github.com/opencv/cvat/tree/develop/tests/python). There are already [several tests](https://github.com/opencv/cvat/blob/d1a300f0189744bef01d2c60135ec474c5004fbc/tests/python/rest_api/test_tasks.py#L2587) for specific cases in dataset uploading, they can be helpful.

Reply:
@zhiltsov-max 
Noted, As of now I am yet to find the cause of issue in the exporting the annotations so I will currently work on the tests for the current issue as requested and once the issue is merged I will continue working on the export error issue. Thank you for the resources I will refer to it and make changes accordingly while creating the test cases!

Reply:
@zhiltsov-max I was able to get the inital tests setup and am able to run the tests sucessfully but using `pytest ./tests/python` and am getting theese results where some test cases are failing as well as status code 500 error which I think is an internal server error and am a little unsure about how to deal with this any pointers would be helpful.

```pytest ==================================================== short test summary info =====================================================
FAILED tests/python/cli/test_cli.py::TestCLI::test_can_upload_annotations - AssertionError
FAILED tests/python/rest_api/test_projects.py::TestImportExportDatasetProject::test_can_import_dataset_in_org - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
FAILED tests/python/rest_api/test_projects.py::TestImportExportDatasetProject::test_can_export_and_import_dataset_with_skeletons_cvat_for_video - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
FAILED tests/python/rest_api/test_projects.py::TestImportExportDatasetProject::test_can_import_export_annotations_with_rotation - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
FAILED tests/python/rest_api/test_tasks.py::TestTaskBackups::test_can_import_backup[interpolation] - AssertionError: assert 'annotation' == 'validation'
FAILED tests/python/rest_api/test_tasks.py::TestUnequalJobs::test_can_import_annotations - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
FAILED tests/python/sdk/test_projects.py::TestProjectUsecases::test_can_create_project_from_dataset - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
ERROR tests/python/cli/test_cli.py::TestCLI::test_can_create_from_backup - cvat_sdk.api_client.exceptions.ServiceException: Status Code: 500
========================== 7 failed, 1299 passed, 38 skipped, 86 warnings, 1 error in 513.60s (0:08:33) ==========================

Reply:
@zhiltsov-max , Additionally I'd appreciate your feedback on my test implementation as I'm relatively new to writing tests. Could you please review it and let me know how I can verify that the test is functioning properly.

```python 
class TestCOCOFormatSameNameImport:
    @staticmethod
    def _make_client() -> Client:
        return Client(BASE_URL, config=Config(status_check_period=0.01))

    @pytest.fixture(
        autouse=True,
        scope="class",
        name="TestCOCOFormatSameNameImport.setup_class",
    )
    @classmethod
    def setup_class(
        cls, restore_db_per_class, tmp_path_factory: pytest.TempPathFactory, admin_user: str
    ):
        cls.tmp_dir = tmp_path_factory.mktemp(cls.__class__.__name__)
        cls.client = cls._make_client()
        cls.user = admin_user
        cls.format_name = "COCO"

        with cls.client:
            cls.client.login((cls.user, USER_PASS))

        cls._init_tasks()

    @classmethod
    def _create_task_with_annotations(cls, filenames: List[str]):
        images = generate_image_files(len(filenames), filenames=filenames)

        source_archive_path = cls.tmp_dir / "source_data.zip"
        with zipfile.ZipFile(source_archive_path, "w") as zip_file:
            for image in images:
                zip_file.writestr(image.name, image.getvalue())

        task = cls.client.tasks.create_from_data(
            {
                "name": "test_images_with_coco_annotations",
                "labels": [{"name": "cat"}, {"name": "dog"}],
            },
            resources=[source_archive_path],
        )

        labels = task.get_labels()
        task.set_annotations(
            models.LabeledDataRequest(
                shapes=[
                    models.LabeledShapeRequest(
                        frame=frame_id,
                        label_id=labels[0].id,
                        type="rectangle",
                        points=[1, 1, 2, 2],
                    )
                    for frame_id in range(len(filenames))
                ],
            )
        )

        return task

    @classmethod
    def _init_tasks(cls):
        filenames = [
            "filename1.jpg",
            "filename1.png",
            "filename2.jpg",
            "filename2.png",
            "filename3.jpg",
            "filename4.png",
        ]

        cls.data = {}
        for filename in filenames:
            task = cls._create_task_with_annotations([filename])
            dataset_file = cls.tmp_dir / f"{filename}_dataset.zip"
            task.export_dataset(cls.format_name, dataset_file, include_images=False)
            cls.data[filename] = (task, dataset_file)

    @pytest.mark.parametrize(
        "filename, expect_success",
        [
            ("filename1.jpg", True),
            ("filename1.png", True),
            ("filename2.jpg", True),
            ("filename2.png", True),
            ("filename3.jpg", True),
            ("filename4.png", True),
        ],
    )
    def test_import_coco_annotations(self, filename, expect_success):
        task: Task = self.data[filename][0]
        dataset_file = self.data[filename][1]

        if expect_success:
            task.import_annotations(self.format_name, dataset_file)
            assert set(s.frame for s in task.get_annotations().shapes) == {0}
        else:
            with pytest.raises(exceptions.ApiException) as capture:
                task.import_annotations(self.format_name, dataset_file)

            assert b"Could not upload annotation for the " in capture.value.body
```
please note that I came up with this refering to the current implementation of the test so do let me know if it is incorrect. Thank you for your support!

Reply:
> I was able to get the inital tests setup and am able to run the tests sucessfully but using pytest ./tests/python and am getting theese results where some test cases are failing as well as status code 500 error which I think is an internal server error and am a little unsure about how to deal with this any pointers would be helpful.

Please try to check server logs with `docker logs test_cvat_server_1`, `docker logs test_cvat_worker_import_1`, `docker logs test_cvat_worker_export_1`. You can also [attach a debugger](https://opencv.github.io/cvat/docs/contributing/running-tests/#rest-api-sdk-and-cli-tests), if it's needed.


> please note that I came up with this refering to the current implementation of the test so do let me know if it is incorrect. Thank you for your support!

- Seems like it does quite a few unnecessary actions. Please consider minimizing the amount of actions done in the test setup.
- Importing the exported dataset makes no sense, as it doesn't reproduce the problem (since the export is broken too wrt. this issue).
- Why the test is checking that there are no annotations in the task after uploading?

Reply:
Hi, any updates?

Reply:
@zhiltsov-max I am sorry for the delayed response and must inform you that progress on this task has been slower than anticipated since I had been a little busy with the proposal submission for GSOC. I plan to complete working on this issue as soon as possible. I apologize for any inconvenience this may have caused, and I assure you that I am dedicated to completing this task promptly and effectively. Thank you for your understanding and continued support. 🙇

Reply:
Ok, do as you see fits best to you, it's your PR after all.